### PR TITLE
Port Staking Widget UI

### LINKED ISSUE #270: Port Staking Widget

### PR DESCRIPTION
**Pull request by willm30**
## Description

**Note**: I recommend testing this pr first: https://github.com/JoinColony/colonyCDapp/pull/299. You will be implicitly testing this pr by testing that one.

This PR ports some of the Staking Widget components (plus a few minor refactors to other components, such as converting to use Hook Form).  Apologies for the large pr. The Staking Widget is small but mighty!

This pr does not wire in the Staking Widget. (https://github.com/JoinColony/colonyCDapp/pull/299)
It also does not port the Objection modal or the grouped staking summary (https://github.com/JoinColony/colonyCDapp/pull/301, https://github.com/JoinColony/colonyCDapp/pull/310)

Also note that this component uses a lot of calculations, mainly derived from motion state and the user's reputation. Where these are shared between subcomponents, I've left them in the StakingWidget provider. Where they are only consumed by one subcomponent, I've put into a component-specific hook.

staking widget

### Setup instructions

1. Build Cdapp on this branch without block ingestor. Run webpack.
2. Checkout and run block ingestor separately on this branch: https://github.com/JoinColony/block-ingestor/pull/18
3. Turn reputation monitor on: http://localhost:3001/reputation/monitor/toggle
4. Give yourself some reputation via truffle console:
    -> c = await IColony.at('{colonyAddress}')
   ->  c.emitDomainReputationReward(1, accounts[0], '10000000000000000000000')
5. Install voting reputation extn
6. Click mint token

You'll be redirected to a rather uninteractive motion staking page. There are some things you can do manually to simulate various different states:

1. In the staking widget provider, change the number of userActivatedTokens (hard-coded for purposes of this pr):
   -> 0 tokens activated
   -> some but not 100% of the tokens needed (e.g '10000000000000000')
2. Log out; then log in with a different user (with a profile), but without reputation

There are other possibilities but I recommend reserving scrutiny for #299. Ultimately, this should be a common-sense check of the ui and code.

It should more or less look like this (only focusing on the staking widget itself): https://xdai.colony.io/colony/willsworld/tx/0xdfd67adb10e6ef997a3316c9b4a95049feb746b60ba0ed11f7cfd60395b5b87d

**New stuff** ✨

* Staking Widget Components

**Changes** 🏗

* Converting components to use hook form, e.g Slider

Resolves #270

### CODE REVIEW

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  },
+  stakeMoreReputation: {
+    id: `${displayName}.stakeMoreReputation`,
+    defaultMessage: `Oops! Your ability to stake is limited by the amount of Reputation you have. To be able to stake more on future motions, you'll need to earn more Reputation.`,
+  },
+  /*
+   * I think this copy is misleading. A motion can be staked more than the user's min stake (obviously).
+   * Shouldn't it be: Your minimum stake is more than the amount needed for the motion to proceed. A motion can't be staked over 100% of the amount needed.' ?
+   */
+  cantStakeMore: {
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage.tsx:38]: I think this copy is misleading. A motion can be staked more than the user's min stake (obviously).
 Shouldn't it be: Your minimum stake is more than the amount needed for the motion to proceed. A motion can't be staked over 100% of the amount needed.' ?

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  //   } = useUserBalanceWithLockQuery({
+  //     variables: {
+  //       address: walletAddress,
+  //       tokenAddress: nativeTokenAddress,
+  //       colonyAddress,
+  //     },
+  //   });
+  //userData?.user?.userLock?.balance || 0,
+
+  const userActivatedTokens = new Decimal('10000000000000000000');
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx:135]: hard-coded for purposes of this pr

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,283 @@
+import React, {
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx:1]: This file contains all of the various calculations depended on by the staking widget and its subcomponents

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  },
+  stakeMoreTokens: {
+    id: `${displayName}.stakeMoreTokens`,
+    defaultMessage: `Oops! You don't have enough active tokens. To stake more than this, please activate more tokens.`,
+  },
+  stakeMoreReputation: {
+    id: `${displayName}.stakeMoreReputation`,
+    defaultMessage: `Oops! Your ability to stake is limited by the amount of Reputation you have. To be able to stake more on future motions, you'll need to earn more Reputation.`,
+  },
+  cantStakeMore: {
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage.tsx:34]: I think this copy is misleading. A motion can be staked more than the user's min stake (obviously).
Shouldn't it be: Your minimum stake is more than the amount needed for the motion to proceed. A motion can't be staked over 100% of the amount needed.' ?

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/StakingSlider.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  'common.ColonyActions.ActionDetailsPage.DefaultMotion.StakingWidget.StakingSlider';
+
+export interface StakingSliderProps {
+  sliderAnnotationProps: SomeSliderAnnotationProps;
+  stakingWidgetSliderProps: SomeStakingWidgetSliderProps;
+  stakingValidationMessageProps: SomeStakingValidationProps;
+  mutableRef?: ReturnType<typeof useRef>;
+}
+
+const StakingSlider = ({
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/StakingSlider.tsx:27]: A note on this component. 

It gets used both in the StakingWidget, and in the RaiseObjection modal (introduced in https://github.com/JoinColony/colonyCDapp/pull/301).

Dialogs don't have access to context unless you wrap the dialog provider with the provider you intend to access from the dialog (which essentially broadcasts to the entire app, like we do with `AppContext`). 

That doesn't feel like the right choice for the `StakingWidgetProvider`. It only needs to wrap the staking widget (not the whole app). So, in order to get all the necessary data to the `StakingSlider` when it's used in the Dialog (and thus outside of the `StakingWidgetProvider`, I'm passing props. A lot of them. 

I've done my best to hide this as much as possible so it doesn't look like hefty prop-drilling is taking place... but it is.

File: src/hooks/motionWidgets/stakingWidget/useStakingInput.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+import { useAppContext, useColonyContext } from '~hooks';
+import { mapPayload } from '~utils/actions';
+
+import { mapStakingSliderProps } from './helpers';
+
+const useStakingInput = () => {
+  const { user } = useAppContext();
+  const { colony } = useColonyContext();
+  const mutableRef = useRef<{ limitExceeded: boolean }>();
```
Comment (willm30) [inline src/hooks/motionWidgets/stakingWidget/useStakingInput.ts:17]: This is used to pass data between siblings in the component tree (using `useImperativeHandle` in `useStakingControls`).

`StakingControls` needs to know if the limit's been exceeded, so it can display the `ActivateButton`, but it doesn't make sense to move the `limitExceeded` state up to a shared parent in this case, because that would obligate us to hold this state in the parent every where the `StakingSlider` is used (e.g. in `ObjectionSlider` also), even if `StakingControls` doesn't also live there (as is the case in the objection modal). 

In other words, the state really does belong inside the `StakingSlider` component, and this allows it to be shared horizontally.

File: amplify/backend/api/colonycdapp/schema.graphql [hunk 7]
```diff
@@ -409,6 +435,7 @@ type ColonyExtension @model {
   isDeleted: Boolean!
   isInitialized: Boolean!
   version: Int!
+  extensionConfig: ExtensionConfiguration
```
Comment (willm30) [inline amplify/backend/api/colonycdapp/schema.graphql:438]: This data is used by the staking widget, so adding here for easy retrieval.

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/MotionPhaseWidget.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+const MotionPhaseWidget = ({
+  actionData: { motionData },
+}: MotionPhaseWidgetProps) => {
+  if (!motionData) {
+    return null;
+  }
+
+  switch (motionData.motionState) {
+    case MotionState.Staking: {
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/MotionPhaseWidget.tsx:22]: To be extended with other widgets

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingInput/StakingControls/ActivateButton.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  'common.ColonyActions.ActionDetailsPage.DefaultMotion.StakingWidget.ActivateButton';
+
+const ActivateButton = () => (
+  <Button
+    appearance={{
+      theme: 'primary',
+      size: 'medium',
+    }}
+    text={{ id: 'button.activate' }}
+    //onClick={() => openTokenActivationPopover(true)}
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingInput/StakingControls/ActivateButton.tsx:15]: https://github.com/JoinColony/colonyCDapp/issues/297

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+  //   } = useUserBalanceWithLockQuery({
+  //     variables: {
+  //       address: walletAddress,
+  //       tokenAddress: nativeTokenAddress,
+  //       colonyAddress,
+  //     },
+  //   });
+  //userData?.user?.userLock?.balance || 0,
+
+  const userActivatedTokens = useMemo(
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx:114]: These `Decimals` are memoised so they don't trigger the a rendering of the entire provider's subtree on each render

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    maxUserStake.lt(remainingToStake.toString());
+
+  /*
+   * Occurs when the remaining amount to be staked is less than the user's minimum stake. (i.e. can't
+   * stake more than 100% of a motion)
+   */
+  const cantStakeMore =
+    remainingToStake.lte(minUserStake) && !remainingToStake.isZero();
+
+  const getErrorType = useCallback(
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx:157]: These are the various states that will need testing

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/helpers.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
+import Decimal from 'decimal.js';
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/helpers.ts:1]: various helpers for calculations

File: src/components/shared/Fields/Form/ActionHookForm.tsx [hunk 13]
```diff
@@ -39,19 +43,23 @@ interface Props<V extends Record<string, any>>
...
   onError,
   onSubmitError,
+  transform,
   submit,
+  error,
   success,
-  transform,
   ...props
 }: Props<V>) => {
+  const submitAction = submit || actionType;
```
Comment (willm30) [inline src/components/shared/Fields/Form/ActionHookForm.tsx:56]: Updating `ActionHookForm` to share same api as `ActionButton` re use of actionType to infer success/error action types

File: src/hooks/motionWidgets/stakingWidget/useRequiredStakeMessage.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
+import Decimal from 'decimal.js';
+import { BigNumber } from 'ethers';
+import { useFormContext } from 'react-hook-form';
+import {
+  convertStakeToPercentage,
+  formatStakePercentage,
+  getStakeFromSlider,
+  SLIDER_AMOUNT,
+} from '~common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget';
```
Comment (willm30) [inline src/hooks/motionWidgets/stakingWidget/useRequiredStakeMessage.ts:9]: I have put all these hooks into our top-level `hooks` directory, since that's the pattern we're using at the moment.

But really, all these hooks could live in their own `hooks` folder alongside the Staking Widget component. This would make the import address shorter, and also seems logical given that these hooks aren't going to be used outside of their respective components.

File: src/hooks/motionWidgets/stakingWidget/useStakingControls.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  } = useStakingWidgetContext();
+  const { getValues } = useFormContext();
+  const sliderAmount: number = getValues(SLIDER_AMOUNT);
+  const stake = getStakeFromSlider(
+    sliderAmount,
+    remainingToStake,
+    minUserStake,
+  );
+
+  const showObjectButton = nayPercentage === '0.00';
```
Comment (willm30) [inline src/hooks/motionWidgets/stakingWidget/useStakingControls.ts:27]: Once you object to a motion, the flow changes so that all future stakes are mediated via the GroupedTotal stake component, thus rendering the object button unnecessary.

File: src/hooks/motionWidgets/stakingWidget/useStakingSlider.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+  mutableRef,
+  remainingToStake,
+  totalPercentage,
+  minUserStake,
+  nativeTokenDecimals,
+  nativeTokenSymbol,
+}: useStakingSliderProps) => {
+  const { user } = useAppContext();
+  const [limitExceeded, setLimitExceeded] = useState(false);
+  useImperativeHandle(mutableRef, () => ({ limitExceeded }), [limitExceeded]);
```
Comment (willm30) [inline src/hooks/motionWidgets/stakingWidget/useStakingSlider.ts:24]: See comment above for explanation

File: src/i18n/en.json [hunk 17]
```diff
@@ -16,32 +16,35 @@
...
-  "button.no": "No",
-  "button.add": "Add",
-  "button.back": "Back",
   "button.close": "Close",
   "button.confirm": "Confirm",
   "button.continue": "Continue",
   "button.createMotion": "Create Motion",
-  "button.next": "Next",
-  "button.save": "Save",
+  "button.decline": "Decline",
```
Comment (willm30) [inline src/i18n/en.json:29]: ordering the `button`s alphabetically 😬

File: src/hooks/motionWidgets/stakingWidget/index.ts [hunk 18]
```diff
@@ -0,0 +1,7 @@
+export { default as useStakingControls } from './useStakingControls';
```
Comment (willm30) [inline src/hooks/motionWidgets/stakingWidget/index.ts:1]: I like this pattern of wrapping all logic into a single hook that exposes only what is consumed by the view its named after. It keeps the view neat.

The only thing I would like to change (per my comment below) is to colocated the hook with its component in the file system, and save the top level "hooks" folder for hooks that get reused around the app. 

👍 or 👎?

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/StakingValidationMessage.tsx [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+  },
+  stakeMoreTokens: {
+    id: `${displayName}.stakeMoreTokens`,
+    defaultMessage: `Oops! You don't have enough active tokens. To stake more than this, please activate more tokens.`,
+  },
+  stakeMoreReputation: {
+    id: `${displayName}.stakeMoreReputation`,
+    defaultMessage: `Oops! Your ability to stake is limited by the amount of Reputation you have. To be able to stake more on future motions, you'll need to earn more Reputation.`,
+  },
+  cantStakeMore: {
```
Comment (willm30) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/StakingValidationMessage.tsx:33]: I think this copy is misleading. A motion can be staked more than the user's min stake (obviously).
Shouldn't it be: Your minimum stake is more than the amount needed for the motion to proceed. A motion can't be staked over 100% of the amount needed.' ?

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/StakingValidationMessage.tsx [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+  limitExceeded: boolean;
+  minUserStake: Decimal;
+  maxUserStake: Decimal;
+  getErrorType: StakingWidgetContextValues['getErrorType'];
+  nativeTokenDecimals: number;
+  nativeTokenSymbol: string;
+  reputationLoading: boolean;
+}
+
+const StakingValidationMessage = ({
```
Comment (rdig) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/StakingValidationMessage.tsx:49]: Not gonna bring it up now, since the UI is going to change, but I'm not really happy with the while motion widgets structure, so I'm hoping we can come up with something better once we get to the redesign

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/StakingValidationMessage.tsx [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    minUserStake,
+    nativeTokenDecimals,
+    nativeTokenSymbol,
+  });
+
+  if (!errorType || reputationLoading) {
+    return null;
+  }
+
+  if (errorType === 'tokens') {
```
Comment (rdig) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/StakingValidationMessage.tsx:74]: Maybe this string can be a type? Or a enum if there are more of them?

_(If `errorType` can only be `tokens`, then just leave it as it is for now)_

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/getErrorValues.tsx [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+          value={minUserStake}
+        />
+      ),
+      userReputation: (
+        <ValidationErrorNumeral
+          decimals={nativeTokenDecimals}
+          value={maxUserStake}
+        />
+      ),
+    },
```
Comment (rdig) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingSlider/SliderMessages/StakingValidationMessage/getErrorValues.tsx:56]: Right. Now I see the error. Yeah, the error types should be `enum` s

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx (hunk 11 above)
Comment (rdig) [reply src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx:159]: These should also be some sort of type, possibly enums

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+      motionId,
+      requiredStake,
+      userStake,
+      nativeTokenDecimals,
+      nativeTokenSymbol,
+      reputationLoading,
+      loadingStakeData,
+      getErrorType,
+    ],
+  );
```
Comment (rdig) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/StakingWidgetProvider.tsx:253]: Holly hell! There must be a better way. Can't we just pass the full motion data to this ? Because in it's current state it's just asking for trouble

File: src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/helpers.ts [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+export const getStakedPercentages = (
+  isObjection: boolean,
+  requiredStake: Decimal,
+  {
+    yay,
+    nay,
+  }: {
+    yay: string;
+    nay: string;
+  },
```
Comment (rdig) [inline src/components/common/ColonyActions/ActionDetailsPage/DefaultMotion/MotionPhaseWidget/StakingWidget/helpers.ts:68]: This is a bit confusing. It's the staked amount? Or the staked side? _(I know it's the amount, but maybe a more descriptive variable name?)_

File: src/graphql/fragments/actions.graphql [hunk 25]
```diff
@@ -23,8 +23,39 @@ fragment ColonyAction on ColonyAction {
...
+  }
+  usersStakes {
+    address
+    stakes {
+      raw {
+        yay
+        nay
+      }
+    }
+  }
```
Comment (rdig) [inline src/graphql/fragments/actions.graphql:55]: These are so confusing. Yeah, after a while I understand what's going on, but just reading them outright is gonna cause trouble for sure.

Problem is I don't have a better naming alternative, without going super descriptive

File: src/hooks/motionWidgets/stakingWidget/helpers.ts [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+  maxUserStake,
+  userActivatedTokens,
+  nativeTokenDecimals,
+  nativeTokenSymbol,
+  totalPercentage,
+  enoughTokens,
+  reputationLoading,
+  getErrorType,
+  mutableRef,
+}): StakingSliderProps => ({
```
Comment (rdig) [inline src/hooks/motionWidgets/stakingWidget/helpers.ts:17]: Same as I mentioned above. Can we pass the whole motion data directly, as destructuring individual keys is going to cause trouble in the future

File: src/hooks/motionWidgets/stakingWidget/index.ts (hunk 18 above)
Comment (rdig) [reply src/hooks/motionWidgets/stakingWidget/index.ts:1]: Sure, we need to refactor hooks anyway (at least their structure), since they've grown exponentially, and it's getting hard to work with them

Just maybe don't do it in this PR, since it's massive as it is already

**Review by rdig — APPROVED** [review]
First pass seems alright, especially for a PR of this size.

I left a bunch of comments all over the place. but I'm also pre-approving this.

**Comment by jakubcolony** [comment]
@willm30 This looks like a rebase is needed

File: amplify/backend/function/updateExtensionByColonyAndHash/src/graphql.js [hunk 27]
```diff
@@ -9,6 +9,10 @@ const extensionFragment = /* GraphQL */ `
     isDeleted
     isInitialized
     version
+    extensionConfig {
+      minimumStake
+      requiredStake
+    }
```
[the remark this answers is not in the document]
Comment (willm30) [reply amplify/backend/function/updateExtensionByColonyAndHash/src/graphql.js:15]: https://github.com/JoinColony/colonyCDapp/issues/327

**Comment by willm30** [comment]
Splitting this pr up since it got a little too big and its treatment of staking widget data needs modifying. Will eventually be closed.

**Comment by willm30** [comment]
Closing in favour of https://github.com/JoinColony/colonyCDapp/pull/329
